WikiWatershed/model-my-watershed
# Model NHD Hi Res Streams

## Overview

Switch to using NHD Hi Res streams for modeling by default. A migration is added to declare all existing projects to NHD Medium Res, so their values should not change. Also switch the Subbasin processing to use them.

Bring in a new version of the geoprocessing service that supports multiple MultiLineStrings for the `/multi` endpoint. See https://github.com/WikiWatershed/mmw-geoprocessing/pull/94

Connects #3429 

### Notes

Discovered #3431 while working on this, will fix that next.

Since the PA DEP Worksheet Export is run _before_ there is a project, it uses the default values, which going forward will be NHD Hi Res streams. There is currently no way to run the Worksheet Export with Medium Res streams. This may be added as a query parameter set by a modal in the future if necessary.

## Testing Instructions

### Preparation

* Start on latest `develop`, then turn on the VMs
    ```console
    vagrant up
    ```
* From LastPass, copy the "New key" value of "SRAT Catchment API MMW (Staging)"
* Use that value as an environment variable to provision the `app` and `worker` VMs
    ```console
    MMW_SRAT_CATCHMENT_API_KEY=*** vagrant provision app worker
    ```

### Baseline

* Go to https://staging.modelmywatershed.org/
* Select a HUC-10 Boundary
* Once Analysis completes, select the Watershed Multi-Year Model
    ![image](https://user-images.githubusercontent.com/1430060/141134935-b7fb3648-7a8a-4587-b2e8-4c3fa1d6dcda.png)
* Once the Watershed Multi-Year Model completes, export the GMS file
    ![image](https://user-images.githubusercontent.com/1430060/141140971-192c3563-eed2-453d-98b6-b8867b8b9e0e.png)
* Rename the file to be `staging-huc10.gms`

### Pre-existing NHD Project

* Go to http://localhost:8000/
* Ensure you are logged in
* Select the same HUC-10 Boundary
* Run the Watershed Multi-Year Model completes, export the GMS file and save it as `local-huc10-nhd-old.gms`
* [x] Diff the two files, ensure there is no difference
    ```console
    diff staging-huc10.gms local-huc10-nhd-old.gms
    ```

### Test NHD Upgrade

* Checkout this branch
* Reprovision worker 
    ```console
    MMW_SRAT_CATCHMENT_API_KEY=*** vagrant provision app worker
    ```
* Run migrations
    ```console
    ./scripts/manage.sh migrate
    ```
* Go to http://localhost:8000/
* Open the project you made in the previous step
* Create a new scenario
    ![image](https://user-images.githubusercontent.com/1430060/141146780-41429f7d-b29c-448f-9e53-1b1c0e0868e2.png)
* Export the GMS file for the new scenario, save it as `local-huc10-nhd-new.gms`
* [x] Diff the old and new files, ensure there is no difference
    ```console
    diff local-huc10-nhd-old.gms local-huc10-nhd-new.gms
    ```

### Test NHD Hi Res Streams

* Go to http://localhost:8000/
* Make a new project with the same HUC-10 shape
* Export the GMS file, save it as `local-huc10-nhdhr.gms`
* [x] Diff with one of the old files. Ensure you see a difference in Line 2, comma positions 9, 10, and 11, which correspond to Total Stream Length, Length of Streams in Agricultural Areas, and Length of Streams in Urban Areas, as described in the [GMS File Key](https://docs.google.com/spreadsheets/d/12hsklRHwl6sNum0xPk6d5Cw6ONDA3X6BR5-Pc-NwJBc)
    ```diff
    diff staging-huc10.gms local-huc10-nhdhr.gms
    2c2
    < 1.4.0,0.03719145907473309,0,10,0,0,0.08913040276368847,8.954994368619333,438209,0,438209,1560.1930842215154,0,6.167108541732857,0.05321929978643071,30,1961,1990,0.002957640540510605,23063.314486478914,0,0,0,0.1816811827761761
    ---
    > 1.4.0,0.03719145907473309,0,10,0,0,0.08913040276368847,8.954994368619333,658483,33674.290302124384,624808.7096978757,1560.1930842215154,0,6.167108541732857,0.05321929978643071,30,1961,1990,0.002957640540510605,23063.314486478914,0,0,0,0.1816811827761761
    ```
* [x] Ensure the new values are greater than the old values

### Test Subbasin

* Switch to the Water Quality tab, then select Subbasin Attenuated Results
    ![image](https://user-images.githubusercontent.com/1430060/141148781-11c1ad37-9804-4149-9c5e-676a30126df2.png)

### Test Worksheet Export

* Go to http://localhost:8000/
* Draw a shape using the draw tool or the 1km square tool
* Select the Watershed Multi-Year Worksheet in the Model tab
    ![image](https://user-images.githubusercontent.com/1430060/141149140-4294edb8-ad47-489c-9066-174383938ab9.png)
